### PR TITLE
Netatmo integration

### PR DESCRIPTION
**Pull request by Terdious**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to the community ([french forum](https://community.gladysassistant.com/)/[english forum](https://en-community.gladysassistant.com/)) for testing before merging.
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Add a new Netatmo integration
OAuth2 authentication management

### CODE REVIEW

**Comment by Terdious** [comment]
Hi @Pierre-Gilles 
Couldn't it be beneficial to do a first review for this part which is new (I want to talk in particular about the OAut2 connection).
This would also prevent the review from being too long.
And also for me not to go work on something else if this start is not valid.

Thanks in advance.

**Comment by Terdious** [comment]
Documentation : https://github.com/GladysAssistant/v4-website/pull/238

**Comment by Terdious** [comment]
Hi @Pierre-Gilles, I'm sorry, but following the update of the branch after your master update, codecov crashed again for nothing. And... I still can't find how to restart the checks... Could you help me? So you don't think the PR build will crash when you can review it...
image

File: front/src/assets/integrations/devices/netatmo/netatmo-NAPlug.jpg
Comment (Pierre-Gilles) [inline front/src/assets/integrations/devices/netatmo/netatmo-NAPlug.jpg:1]: This image is too big :) You need to resize+compress it. This kind of image could be easily 50ko

File: front/src/assets/integrations/devices/netatmo/netatmo-NATherm1.jpg
Comment (Pierre-Gilles) [inline front/src/assets/integrations/devices/netatmo/netatmo-NATherm1.jpg:1]: This image is too big :) You need to resize+compress it. This kind of image could be easily 50ko

File: front/src/routes/integration/all/netatmo/device-page/index.js
Comment (Pierre-Gilles) [inline front/src/routes/integration/all/netatmo/device-page/index.js]: Unused code should be removed

File: front/src/routes/integration/all/netatmo/setup-page/style.css
Comment (Pierre-Gilles) [inline front/src/routes/integration/all/netatmo/setup-page/style.css]: french comment, should be in english

File: server/services/netatmo/api/netatmo.controller.js
Comment (Pierre-Gilles) [inline server/services/netatmo/api/netatmo.controller.js]: our API is snake_case + following the REST principle ( https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api ) 

In that case, it should be: 

```
POST /api/v1/service/netatmo/configuration
```

File: server/services/netatmo/api/netatmo.controller.js
Comment (Pierre-Gilles) [inline server/services/netatmo/api/netatmo.controller.js]: Same here: 

```
POST /api/v1/service/netatmo/status
```

File: server/services/netatmo/api/netatmo.controller.js
Comment (Pierre-Gilles) [inline server/services/netatmo/api/netatmo.controller.js]: Same here, if you do a get, you should do: 

```
GET /api/v1/service/netatmo/token
```

File: server/services/netatmo/api/netatmo.controller.js
Comment (Pierre-Gilles) [inline server/services/netatmo/api/netatmo.controller.js]: why wrap the result in a "result" attribute and not just return the object? Avoid un-necessary wrappers :)

File: server/services/netatmo/api/netatmo.controller.js
Comment (Pierre-Gilles) [inline server/services/netatmo/api/netatmo.controller.js]: Same here, avoid `refreshDiscover`.

Maybe: 

```
GET /api/v1/service/netatmo/device
```

?

File: server/services/netatmo/lib/device/examples.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/device/examples.js]: Old file?

File: server/services/netatmo/lib/index.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/index.js]: Wrong comment?

File: server/services/netatmo/lib/index.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/index.js]: could this function be moved to another file?

File: server/services/netatmo/lib/netatmo.connect.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.connect.js]: Why wrap this variable in a string with no other content around? Is it to convert it to String?

File: server/services/netatmo/lib/netatmo.discoverDevices.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.discoverDevices.js]: why do you pass the current object as param of the function? You are inside an object so you have access to `this`

File: server/services/netatmo/lib/netatmo.getConfiguration.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.getConfiguration.js]: Same

File: server/services/netatmo/lib/netatmo.getTokens.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.getTokens.js]: same

File: server/services/netatmo/lib/netatmo.getTokens.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.getTokens.js]: Can you separate this file into 2 files? 

1 function = 1 file

File: server/services/netatmo/lib/netatmo.saveValue.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.saveValue.js]: Bad comment?

File: front/src/routes/integration/all/netatmo/NetatmoDeviceBox.jsx
Comment (Pierre-Gilles) [inline front/src/routes/integration/all/netatmo/NetatmoDeviceBox.jsx]: Be careful with relative path, in other parts of the code we use absolute path

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thanks for this PR 🙏 

I haven't entirely tested the integration, I just did a first review with what I could test on the UI side + a first code review.

**On the functional side**

When I first open the integration, here is what I see:

Screenshot 2024-01-11 at 09 54 11

There is an error message immediately (I didn't do anything), which is a bit weird for the user.

Link to the documentation is missing category: 
Screenshot 2024-01-11 at 09 54 30

Buttons a bit weird:

Screenshot 2024-01-11 at 09 55 08

**On the code side**

The code is so much different from when I first had a look 😅

- There is a major issue with the way the `NetatmoHandler` works, it seems you are passing the object (itself) to every single function. It's an anti-pattern, you have access to `this` everywhere. 
- Many files of this class contains several functions. Try to keep 1 file = 1 function when you are building functions of a class (Talking about NetatmoHandler)
- Some function comments are not explaining what a function does, making the code hard to grasp.
- See all my comments

Please take some time to review the PR entirely and improve those points before I can do a second review. 

This is already a massive PR (+7000 lines of code) so each review is expensive.

File: front/src/routes/integration/all/netatmo/NetatmoDeviceBox.jsx
Comment (Terdious) [reply front/src/routes/integration/all/netatmo/NetatmoDeviceBox.jsx]: Ok, corrected

**Comment by Terdious** [comment]
Hi @Pierre-Gilles 

> When I first open the integration, here is what I see:
> 
> Screenshot 2024-01-11 at 09 54 11
> 
> There is an error message immediately (I didn't do anything), which is a bit weird for the user.

So for now I can't even reproduce your first connection error... Nobody has obviously had the case on the 2 testers and I have just started again with a fresh installation on wsl, I don't have this problem ..., my state first connection :
image

Also test via the docker image, my state first connection : 
image
Could you see on your side to guide me?

> not very clean, we don't put error message here, it's a menu
> image

Ok for the message part in the left banner, it allowed us to have a common visual and removed the repetition aspect and it seems to me that the 2 testers found it clean on the contrary. I withdraw.

> Link to the documentation is missing category:
> Screenshot 2024-01-11 at 09 54 30

Corrected, my bad !

> Buttons a bit weird:
> 
> Screenshot 2024-01-11 at 09 55 08

Yes for the button, I wanted things to be clearly expressed. This is not just a matter of saving the modified data, but of having to request new access and therefore go through the Netatmo login page each time you save. I make a proposal in this sense to put a text above. Corrected

> There is a major issue with the way the NetatmoHandler works, it seems you are passing the object (itself) to every single function. It's an anti-pattern, you have access to this everywhere.

For the "netatmoHandler" passed around... I guess it's a poor understanding/configuration of my vscode. The “this.” » is declared in eslint error everywhere in my vscode interface, so I passed the 'netatmoHandler' for this reason only. My bad, corrected

> Many files of this class contains several functions. Try to keep 1 file = 1 function when you are building functions of a class (Talking about NetatmoHandler)

I understand. Corrected

File: server/services/netatmo/api/netatmo.controller.js
Comment (Terdious) [reply server/services/netatmo/api/netatmo.controller.js]: It's well POST here but ok for "POST /api/v1/service/netatmo/token"

File: server/services/netatmo/lib/netatmo.connect.js
Comment (Terdious) [reply server/services/netatmo/lib/netatmo.connect.js]: My bad

File: server/services/netatmo/lib/index.js
Comment (Terdious) [reply server/services/netatmo/lib/index.js]: Ok

File: server/services/netatmo/lib/netatmo.loadDevices.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.loadDevices.js]: You are doing a Promise.all on all devices ? It seems that each loadDeviceDetails call the API, so if the user has 50 Netatmo devices, it'll start 50 API calls in parallel, it's way to heavy!

Consider using bluebird Promise.map instead so you can specify a max concurrency :) 

```
const results = await Promise.map(homes, async(home) => 
... 
}, {concurrency: 2} );
```

File: server/services/netatmo/lib/netatmo.loadDeviceDetails.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.loadDeviceDetails.js]: I don't find any async code in this function? If it's sync, use just .map instead of Promise.all

File: server/services/netatmo/lib/netatmo.updateValues.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.updateValues.js]: Weird pattern, why bind "thi"s instead of just having the function in the netatmo object if you need "this"? 

If you don't need this, just build it as a helper function

File: server/services/netatmo/lib/netatmo.updateValues.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.updateValues.js]: Same

File: server/services/netatmo/lib/netatmo.pollRefreshingValues.js
Comment (Pierre-Gilles) [inline server/services/netatmo/lib/netatmo.pollRefreshingValues.js]: You cannot do async work in a .map (sync code)

You need to use bluebird Promise.map instead. Don't forget to set a concurrency

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thanks for the fixes! 🙏

It's way better now :) 

I have a few feedbacks on the server side but nothing serious.

On the UI, it looks good to me now! Can you update your documentation PR with new screenshots?

Once everything is fixed, it'll good for merge I think !! 🚀🚀🚀

File: server/services/netatmo/lib/netatmo.loadDevices.js
Comment (Terdious) [reply server/services/netatmo/lib/netatmo.loadDevices.js]: I'm understand, There is normally not too much risk in this regard because the requests are unique for each house. It's very unlikely that a user will have more than 5 houses so I didn't do too much research. But this way it's much cleaner. THANKS.

File: server/services/netatmo/lib/netatmo.loadDeviceDetails.js
Comment (Terdious) [reply server/services/netatmo/lib/netatmo.loadDeviceDetails.js]: Indeed !!

File: server/services/netatmo/lib/netatmo.loadDevices.js
Comment (Terdious) [reply server/services/netatmo/lib/netatmo.loadDevices.js:29]: I'm understand, There is normally not too much risk in this regard because the requests are unique for each house. It's very unlikely that a user will have more than 5 houses so I didn't do too much research. But this way it's much cleaner. THANKS.

File: server/services/netatmo/lib/netatmo.pollRefreshingValues.js
Comment (Terdious) [reply server/services/netatmo/lib/netatmo.pollRefreshingValues.js]: How to evaluate the competition? Is 10 really suitable? I admit I don't really know. In any case I currently have no lag

File: server/services/netatmo/lib/netatmo.updateValues.js
Comment (Terdious) [reply server/services/netatmo/lib/netatmo.updateValues.js]: Indeed, it simplifies... I assumed it would clutter the netatmo object... Sorry.

**Comment by Terdious** [comment]
> On the UI, it looks good to me now! Can you update your documentation PR with new screenshots?

I do this before the end of the evening.

File: server/services/netatmo/lib/netatmo.pollRefreshingValues.js
Comment (Pierre-Gilles) [reply server/services/netatmo/lib/netatmo.pollRefreshingValues.js]: I would recommended maximum of 2 for something background like this.

**Review by Pierre-Gilles — APPROVED** [review]
Good for me for this first PR, thanks for this great work 👏👏👏

**Pull request merged**
